execution/streaming-partitioner/src/transaction_graph_partitioner.rs

```diff
@@ -0,0 +1,611 @@
+        let graph_stream = TransactionGraphStream::new(transactions, self.params.clone());
+
+        // partitioned_graph_stream: BatchedStream<StreamItem = (SerializationIdx, PartitionId)>
+        let partitioned_graph_stream = self.graph_partitioner.partition_stream(graph_stream)?;
+
+        Ok(
+            partitioned_graph_stream.map_items(|(node, partition)| PartitionedTransaction {
+                transaction: node.data.transaction,
+                serialization_idx: node.index as SerializationIdx,
+                partition,
+                dependencies: node.data.dependencies,
+            }),
+        )
+    }
+}
+
+/// Takes a stream of transactions and turns it into a stream of the dependency graph.
+pub struct TransactionGraphStream<S, NF, EF>
+where
+    S: BatchedStream,
+    S::StreamItem: PTransaction,
+{
+    transactions: S,
+    params: Params<NF, EF>,
+    next_serialization_idx: SerializationIdx,
+    last_write: HashMap<<S::StreamItem as PTransaction>::Key, SerializationIdx>,
+    partitioned_node_weight: NodeWeight,
+    partitioned_edge_weight: EdgeWeight,
+    edges: Vec<Vec<(NodeIndex, EdgeWeight)>>,
+    rng: rand::rngs::ThreadRng,
+}
+
+impl<T, S, NF, EF> TransactionGraphStream<S, NF, EF>
+where
+    T: PTransaction,
+    T::Key: Clone + Eq + std::hash::Hash,
+    S: BatchedStream<StreamItem = T>,
+    NF: Clone + Fn(&T) -> NodeWeight,
+    EF: Clone + Fn(SerializationIdx, SerializationIdx) -> EdgeWeight,
+{
+    fn new(transactions: S, params: Params<NF, EF>) -> Self {
+        Self {
+            transactions,
+            params,
+            next_serialization_idx: 0,
+            last_write: HashMap::new(),
+            partitioned_node_weight: 0 as NodeWeight,
+            partitioned_edge_weight: 0 as EdgeWeight,
+            edges: Vec::new(),
+            rng: rand::thread_rng(), // TODO: allow using a custom RNG.
+        }
+    }
+
+    fn add_transaction(&mut self, tx: T) -> StreamNode<Self> {
+        let idx = self.next_serialization_idx;
+        assert_eq!(idx as usize, self.edges.len());
+        self.next_serialization_idx += 1;
+        self.edges.push(Vec::new());
+
+        // Find this transaction's dependencies.
+        let deps: HashMap<SerializationIdx, Vec<T::Key>> = tx
+            .read_set()
+            .filter_map(|key| Some((*self.last_write.get(key)?, key.clone())))
+            .into_group_map();
+
+        let mut new_edges_weight = 0 as EdgeWeight;
+
+        // Add edges to the dependency graph based on the dependencies.
+        let basic_network_latency_penalty: EdgeWeight = 10;
+        let cross_round_discount: EdgeWeight = 10;
+        let per_key_penalty: EdgeWeight = 1;
+        let idx_round = idx / 48;
+        for (&dep, keys) in deps.iter() {
+            let dep_round = dep / 48;
+            let edge_weight = per_key_penalty * (keys.len() as EdgeWeight) + min(0, basic_network_latency_penalty - cross_round_discount * ((idx_round - dep_round) as EdgeWeight));
```

Comment:
Did you mean `max` instead of `min`? Also, with `basic_network_latency_penalty == cross_round_discount`, will it be just binary? I mean, the second part will be either 0 or 10, right?

In any case, only experiments will show which edge weight function to use, it's very hard to estimate costs theoretically.

Reply:
Did mean `max`... thanks for catching!

In my very rough estimation, it currently takes up to 2ms for a txn to move from ready state to committed state in blockstm, which can be higher than the intrinsic network latency. Reordering may matter more.

Anyway, totally agree, will continue tuning and let the benchmark number drive.

And thanks for your great job!